README
PHP client library for Okuyama(Distributed key-value-store).

System requirement
  PHP 5.3
  PHPUnit

Okuyama supports original protocol, memcached protocol.
This client supports original protocol.
  Net/Okuyama/Adapter/Socket.php
memcached protocol is not support yet.

Okuyama has default PHP client library but it's implementaion is not a PHP way.
So I implement new client library.


CHANGELOG
2011-02-11  Shinya Ohyanagi <sohynagi@gmail.com>

	* Add isConnected() method for get connection alive.


tests/prepare.php
<?php
/**
 * Scripts for PHPUnit
 *
 * PHP version 5.3
 *
 * Copyright (c) 2010-2011 Shinya Ohyanagi, All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *   * Neither the name of Shinya Ohyanagi nor the names of his
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * @category  Net
 * @package   Net\Okuyama
 * @version   $id$
 * @copyright (c) 2010-2011 Shinya Ohyanagi
 * @author    Shinya Ohyanagi <sohyanagi@gmail.com>
 * @license   New BSD License
 */

error_reporting(E_ALL | E_STRICT);
$src = dirname(__DIR__) . DIRECTORY_SEPARATOR . 'src';
set_include_path(get_include_path()
    . PATH_SEPARATOR . $src
);

require_once 'PHPUnit/Framework/TestCase.php';


tests/Okuyama/BasicTest.php
<?php
/**
 * Spec of \Net\Okuyama.
 *
 * PHP version 5.3
 *
 * Copyright (c) 2011 Shinya Ohyanagi, All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *   * Neither the name of Shinya Ohyanagi nor the names of his
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * @use       \Net
 * @category  \Net
 * @package   \Net\Okuyama
 * @version   $id$
 * @copyright (c) 2011 Shinya Ohyanagi
 * @author    Shinya Ohyanagi <sohyanagi@gmail.com>
 * @license   New BSD License
 */

namespace Net;
use Net;

/**
 * @see prepare
 */
require_once dirname(__DIR__) . '/prepare.php';

/**
 * @see \Net\Okuyama
 */
require_once 'Net/Okuyama.php';

/**
 * Basic test.
 *
 * <pre>
 *   Before you run this test, make sure okuyama running.
 * </pre>
 *
 * @use       \Net
 * @category  \Net
 * @package   \Net\Okuyama
 * @version   $id$
 * @copyright (c) 2011 Shinya Ohyanagi
 * @author    Shinya Ohyanagi <sohyanagi@gmail.com>
 * @license   New BSD License
 */
class BasicTest extends \PHPUnit_Framework_TestCase
{
    const KEY_PREFIX = 'Net_Okuyama_Test_';

    private $_client = null;

    private $_hosts = array(
        '127.0.0.1:8888'
    );

    public function setUp()
    {
        $this->_client = new Okuyama();
    }

    public function testShouldCreateInstance()
    {
        $this->assertTrue($this->_client instanceof Okuyama);
    }

    public function testshouldConnetToHost()
    {
        $result = $this->_client->connect($this->_hosts);
        $this->assertTrue($result instanceof \Net\Okuyama);
    }

    public function testShouldCloseConnection()
    {
        $result = $this->_client->connect($this->_hosts);
        $this->assertTrue($this->_client->close());
    }

    /**
     * @expectedException \Net\Okuyama\Exception
     */
    public function testShouldThrowsExceptionWhenConnctionCloseFailed()
    {
        $this->_client->close();
    }

    public function testShouldReturnTrueWhenConnected()
    {
        $this->_client->connect($this->_hosts);
        $ret = $this->_client->isConnected();
        $this->assertTrue($ret);
        $this->_client->close();
    }

    public function testShouldReturnFalseWhenNotConnected()
    {
        $ret = $this->_client->isConnected();
        $this->assertFalse($ret);
    }

    public function testShouldConnectToHostAuto()
    {
        $this->_hosts[] = '127.0.0.1:8889';
        $result = $this->_client->connect($this->_hosts);
        $this->assertTrue($result instanceof \Net\Okuyama);
        $this->_client->close();
    }

    public function testShouldSetDataToServer()
    {
        $this->_client->connect($this->_hosts);
        $ret = $this->_client->set(self::KEY_PREFIX . 'foo', 'bar');
        $this->assertTrue($ret instanceof \Net\Okuyama);

        $value = $this->_client->get(self::KEY_PREFIX . 'foo');
        $this->assertSame($value, 'bar');
        $this->_client->close();
    }

    public function testShouldSetDataWithTag()
    {
        $this->_client->connect($this->_hosts);
        $tags = array('fiz', 'baz');
        $this->_client->set(self::KEY_PREFIX . 'bar', 'foo', $tags)
                      ->set(self::KEY_PREFIX . 'hoge', 'fuga', $tags)
                      ->set(self::KEY_PREFIX . 'foo', 'bar', $tags);

        $result = $this->_client->getKeysByTag('fiz');
        $this->assertSame($result, array(
            self::KEY_PREFIX . 'bar',
            self::KEY_PREFIX . 'hoge',
            self::KEY_PREFIX . 'foo'
        ));
        $this->_client->close();
    }

    public function testShouldGetDataWithTag()
    {
        $this->_client->connect($this->_hosts);
        $tags = array('fiz', 'baz');
        $this->_client->set(self::KEY_PREFIX . 'bar', 'foo', $tags)
                      ->set(self::KEY_PREFIX . 'hoge', 'fuga', $tags)
                      ->set(self::KEY_PREFIX . 'foo', 'bar', $tags);
        $result = $this->_client->getKeysByTag('baz');
        $this->assertSame($result, array(
            self::KEY_PREFIX . 'bar',
            self::KEY_PREFIX . 'hoge',
            self::KEY_PREFIX . 'foo'
        ));
        $this->_client->close();
    }

    public function testShouldReturnNullWhenTagNotExitsts()
    {
        $this->_client->connect($this->_hosts);
        $tags = array('fiz', 'baz');
        $this->_client->set(self::KEY_PREFIX . 'bar', 'foo', $tags)
                      ->set(self::KEY_PREFIX . 'hoge', 'fuga', $tags)
                      ->set(self::KEY_PREFIX . 'foo', 'bar', $tags);

        $result = $this->_client->getKeysByTag('baz' . rand(0, 1000));
        $this->assertSame($result, null);
        $this->_client->close();
    }

    public function testShouldReturnNullWhenArgReturnsSetFalse()
    {
        $this->_client->connect($this->_hosts);
        $tags = array('fiz', 'baz');
        $this->_client->set(self::KEY_PREFIX . 'bar', 'foo', $tags)
                      ->set(self::KEY_PREFIX . 'hoge', 'fuga', $tags)
                      ->set(self::KEY_PREFIX . 'foo', 'bar', $tags);

        $this->_client->remove(self::KEY_PREFIX . 'bar');
        $this->_client->remove(self::KEY_PREFIX . 'hoge');
        $this->_client->remove(self::KEY_PREFIX . 'foo');
        $ret = $this->_client->get(self::KEY_PREFIX . 'bar');

        $result = $this->_client->getKeysByTag('baz', false);
        $this->assertSame($result, null);
        $this->_client->close();
    }

    public function testShouldReturnKeyWhenArgReturnsSetTrue()
    {
        $this->_client->connect($this->_hosts);
        $tags = array('fiz', 'baz');
        $this->_client->set(self::KEY_PREFIX . 'bar', 'foo', $tags)
                      ->set(self::KEY_PREFIX . 'hoge', 'fuga', $tags)
                      ->set(self::KEY_PREFIX . 'foo', 'bar', $tags);

        $this->_client->remove(self::KEY_PREFIX . 'bar');
        $this->_client->remove(self::KEY_PREFIX . 'hoge');
        $this->_client->remove(self::KEY_PREFIX . 'foo');

        $result = $this->_client->getKeysByTag('fiz', true);
        $this->assertSame($result, array(
            self::KEY_PREFIX . 'bar',
            self::KEY_PREFIX . 'hoge',
            self::KEY_PREFIX . 'foo'
        ));
        $this->_client->close();
    }

    public function testShouldReturnTrueWhenDeleteSuccess()
    {
        $this->_client->connect($this->_hosts);
        $this->_client->set(self::KEY_PREFIX . 'foo', 'bar');
        $ret = $this->_client->remove(self::KEY_PREFIX . 'foo');
        $this->assertTrue($ret);

        $value = $this->_client->get(self::KEY_PREFIX . 'foo');
        $this->assertSame($value, null);
        $this->_client->close();
    }

    public function testShouldReturnFalseWhenRemoveFalse()
    {
        $this->_client->connect($this->_hosts);
        $this->_client->set(self::KEY_PREFIX . 'foo', 'bar');
        $ret = $this->_client->remove(self::KEY_PREFIX . 'foo');
        $this->assertTrue($ret);

        $ret = $this->_client->remove(self::KEY_PREFIX . 'foo');
        $this->assertFalse($ret);
        $this->_client->close();
    }

    public function testShouldAddData()
    {
        $this->_client->connect($this->_hosts);
        $ret = $this->_client->remove(self::KEY_PREFIX . 'baz');
        $ret = $this->_client->add(self::KEY_PREFIX . 'baz', 'fiz');
        $this->assertTrue($ret instanceof \Net\Okuyama);

        $this->_client->close();
    }

    public function testShouldReturnFalseWhenDataAlreadyAdded()
    {
        $this->_client->connect($this->_hosts);
        $ret = $this->_client->remove(self::KEY_PREFIX . 'baz');
        $this->_client->add(self::KEY_PREFIX . 'baz', 'fiz');
        $ret = $this->_client->add(self::KEY_PREFIX . 'baz', 'fiz');

        $this->assertTrue($ret instanceof \Net\Okuyama);
        $this->_client->close();
    }

    public function testShouldSetDataWithUsingSetKeyPrefix()
    {
        $this->_client->connect($this->_hosts);
        $this->_client->setKeyPrefix(self::KEY_PREFIX);

        $tags = array('fiz', 'baz');
        $this->_client->set('bar', 'foo', $tags)
                      ->set('hoge', 'fuga', $tags)
                      ->set('foo', 'bar', $tags);

        $result = $this->_client->getKeysByTag('fiz', true);
        $this->assertSame($result, array(
            self::KEY_PREFIX . 'bar',
            self::KEY_PREFIX . 'hoge',
            self::KEY_PREFIX . 'foo'
        ));

        $this->_client->setKeyPrefix('');
        $this->assertSame($this->_client->get(self::KEY_PREFIX . 'bar'), 'foo');
        $this->assertSame($this->_client->get(self::KEY_PREFIX . 'hoge'), 'fuga');
        $this->assertSame($this->_client->get(self::KEY_PREFIX . 'foo'), 'bar');


        $this->_client->setKeyPrefix(self::KEY_PREFIX);
        $this->assertTrue($this->_client->remove('bar'));
        $this->assertTrue($this->_client->remove('hoge'));
        $this->assertTrue($this->_client->remove('foo'));

        $this->_client->close();
    }

    public function testShouldGetVersionNo()
    {
        $this->_client->connect($this->_hosts);
        $this->_client->setKeyPrefix(self::KEY_PREFIX);
        $this->_client->remove('foo');
        $this->_client->set('foo', 'bar');

        $ret = $this->_client->gets('foo');
        $this->assertSame($ret, array('value' => 'bar', 'version' => 0));

        $this->_client->close();
    }

    public function testShouldCheackAndVersion()
    {
        $this->_client->connect($this->_hosts);
        $this->_client->setKeyPrefix(self::KEY_PREFIX);
        $this->_client->remove('foo');
        $this->_client->set('foo', 'bar');

        $ret = $this->_client->gets('foo');
        $this->_client->cas('foo', 'foo', $ret['version']);
        $ret = $this->_client->gets('foo');
        $this->assertSame($ret, array('value' => 'foo', 'version' => 1));
        $this->_client->close();
    }

    public function testShouldSetFailedWhenSetOldVersionNo()
    {
        $this->_client->connect($this->_hosts);
        $this->_client->setKeyPrefix(self::KEY_PREFIX);
        $this->_client->remove('foo');
        $this->_client->set('foo', 'bar');

        $ret = $this->_client->gets('foo');
        $this->_client->cas('foo', 'foo', $ret['version']);
        $ret = $this->_client->gets('foo');
        $this->assertSame($ret, array('value' => 'foo', 'version' => 1));

        $ret = $this->_client->cas('foo', 'baz', 0);
        $this->assertFalse($ret);
        $ret = $this->_client->gets('foo');
        $this->assertSame($ret, array('value' => 'foo', 'version' => 1));
        $this->_client->close();
    }

    public function testShouldRunJavaScriptCode()
    {
        $this->_client->connect($this->_hosts);
        $this->_client->set(self::KEY_PREFIX . 'foo', 'bar');
        $script = <<<EOT
var dataValue;
var retValue = 'foo' + dataValue;
var execRet = '1';
EOT;
        $ret = $this->_client->playScript(self::KEY_PREFIX . 'foo', $script);
        $this->assertSame($ret, 'foobar');
        $ret = $this->_client->get(self::KEY_PREFIX . 'foo');
        $this->assertSame($ret, 'bar');

        $this->_client->remove(self::KEY_PREFIX . 'foo');
        $this->_client->close();
    }

    public function testShouldRunJavaScriptCodeAndUpdateValue()
    {
        $this->_client->connect($this->_hosts);
        $this->_client->set(self::KEY_PREFIX . 'foo', 'bar');
        $script = <<<EOT
var dataValue;
var retValue = 'foo' + dataValue;
var execRet = '2';
EOT;
        $ret = $this->_client->playScript(self::KEY_PREFIX . 'foo', $script, true);
        $this->assertSame($ret, 'foobar');
        $ret = $this->_client->get(self::KEY_PREFIX . 'foo');
        $this->assertSame($ret, 'foobar');

        $this->_client->remove(self::KEY_PREFIX . 'foo');
        $this->_client->close();
    }
}


src/Net/Okuyama.php
<?php
/**
 * Okuyama(Distributed key-value-store) client library.
 *
 * PHP version 5.3
 *
 * Copyright (c) 2011 Shinya Ohyanagi, All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *   * Neither the name of Shinya Ohyanagi nor the names of his
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * @use       \Net
 * @category  \Net
 * @package   \Net\Okuyama
 * @version   $id$
 * @copyright (c) 2011 Shinya Ohyanagi
 * @author    Shinya Ohyanagi <sohyanagi@gmail.com>
 * @license   New BSD License
 */

namespace Net;
use Net,
    Net\Okuyama\Exception;

/**
 * Okuyama(Distributed key-value-store) client library.
 *
 * @use       \Net
 * @category  \Net
 * @package   \Net\KyotoTycoon
 * @version   $id$
 * @copyright (c) 2011 Shinya Ohyanagi
 * @author    Shinya Ohyanagi <sohyanagi@gmail.com>
 * @license   New BSD License
 */
class Okuyama
{
    /**
     * Version.
     */
    const VERSION = '0.0.2';

    /**
     * Client to access Okuyama.
     *
     * @var    mixed
     * @access protected
     */
    protected $_client = null;

    /**
     * Adapter class.
     *
     * @var    mixed
     * @access protected
     */
    protected $_adapter = '\Net\Okuyama\Adapter\Socket';

    /**
     * Key prefix.
     *
     * @var    string
     * @access protected
     */
    protected $_keyPrefix = '';

    /**
     * Get key prefix.
     *
     * @access public
     * @return string Key prefix
     */
    public function getKeyPrefix()
    {
        return $this->_keyPrefix;
    }

    /**
     * Set key prefix.
     *
     * @param  mixed $prefix
     * @access public
     * @return \Net\Okuyama Fluent interface
     */
    public function setKeyPrefix($prefix)
    {
        $this->_keyPrefix = $prefix;
        return $this;
    }

    /**
     * Create client instance.
     *
     * @param  mixed $adapter
     * @access public
     * @return Net\Okuyama Fluent interface
     */
    public function createInstance($adapter)
    {
        // TODO: Use ReflectionClass.
        $this->_client = new $adapter;
        return $this;
    }

    /**
     * Set config options.
     *
     * @param  array $configs
     * @access public
     * @return \Net\Okuyama Fluent interface
     */
    public function setConfig(array $configs)
    {
        if ($this->_adapter === null) {
            throw new Exception('Adapter does not set.');
        }
        $this->_client->setConfig($configs);

        return $this;
    }

    /**
     * Constructor
     *
     * <pre>
     *   array(
     *     'adapter' => '\Net\Okuyama\Adapter\Socket', // Adapter class name.
     *     'timeout' => 10 // Timeout second.
     *   );
     * </pre>
     *
     * @param  array $configs
     * @access public
     * @return void
     */
    public function __construct(array $configs = array())
    {
        spl_autoload_register(array(__CLASS__, 'autoload'));
        if (isset($configs['adapter'])) {
            $this->_adapter = $configs['adapter'];
        }

        $this->createInstance($this->_adapter);
        if ($configs !== array()) {
            $this->setConfig($configs);
        }
    }

    /**
     * Connect to server.
     *
     * @param  array $hosts
     * @access public
     * @return \Net\Okuyama Fluent interface
     */
    public function connect(array $hosts)
    {
        $ret = $this->_client->autoConnect($hosts);
        if ($ret === false) {
            throw new Exception('Connection refused.');
        }

        return $this;
    }

    /**
     * Close connection.
     *
     * @access public
     * @return bool true: Success to close
     */
    public function close()
    {
        return $this->_client->close();
    }

    /**
     * Is connected?
     *
     * @access public
     * @return bool true: Connected, false: Not connected
     */
    public function isConnected()
    {
        return $this->_client->isConnected();
    }

    /**
     * Retrieve the value of a record.
     *
     * @param  mixed $key Key string
     * @param  mixed $tag Tag string
     * @access public
     * @return mixed Value.
     */
    public function get($key)
    {
        return $this->_client->get($this->_keyPrefix . $key);
    }

    /**
     * Set the value of a record.
     *
     * @param  mixed $key The key of the record
     * @param  mixed $value The value of the record
     * @param  mixed $tag Tags
     * @access public
     * @return \Net\Okuyama Fluent interface
     */
    public function set($key, $value, array $tags = array())
    {
        $ret = $this->_client->set($this->_keyPrefix . $key, $value, $tags);
        return $this;
    }

    /**
     * Add a record.
     *
     * @param  mixed $key The key of the record
     * @param  mixed $value The value of the record
     * @param  mixed $tag Tags
     * @access public
     * @return \Net\Okuyama Fluent interface
     */
    public function add($key, $value, array $tags = array())
    {
        $ret = $this->_client->add($this->_keyPrefix . $key, $value, $tags);

        return $this;
    }

    /**
     * Get version no and value at once.
     *
     * <pre>
     *   Equivalent to Memcached gets.
     * </pre>
     *
     * @param  mixed $key The key of the record
     * @access public
     * @return array The value of the record and version no
     */
    public function gets($key)
    {
        return $this->_client->gets($this->_keyPrefix . $key);
    }

    /**
     * Check and set record.
     *
     * <pre>
     *   Equivalent to Memcached cas.
     * </pre>
     *
     * @param  mixed $key The key of the record
     * @param  mixed $value The value of the record
     * @param  array $tags Tags
     * @param  mixed $version Version no
     * @access public
     * @return mixed Fluent interface or false
     */
    public function cas($key, $value, $version, array $tags = array())
    {
        return $this->_client->cas($this->_keyPrefix . $key, $value, $version, $tags);
    }

    /**
     * Remove a record.
     *
     * @param  mixed $key The key of the record
     * @access public
     * @return mixed Result of remove command
     */
    public function remove($key)
    {
        return $this->_client->remove($this->_keyPrefix . $key);
    }

    /**
     * Get keys by tag.
     *
     * @param  mixed $tag Tags
     * @param  mixed $returns
     * @access public
     * @return mixed Keys or null
     */
    public function getKeysByTag($tag, $returns = false)
    {
        return $this->_client->getKeysByTag($tag, $returns);
    }

    /**
     * Retrieve records at once
     *
     * @param  array $key The key of the record
     * @access public
     * @return array Values
     */
    public function getBulk(array $keys)
    {
        $result = array();
        foreach ($keys as $key) {
            $result[$key] = $this->_client->get($this->_keyPrefix . $key);
        }

        return $result;
    }

    /**
     * Set record at once.
     *
     * @param  array $vals
     * @access public
     * @return \Net\Okuyama Fluent interface
     */
    public function setBulk(array $vals)
    {
        foreach ($vals as $v) {
            if (isset($v['key']) && isset($v['value'])) {
                $tags = (isset($v['tags']) && is_array($v['tags'])) ? $v['tags'] : null;

                $this->_client->set(
                    $this->_keyPrefix . $v['key'], $v['value'], $tags
                );
            }
        }

        return $this;
    }

    /**
     * Remove at onece.
     *
     * @param  array $key The key of the record
     * @access public
     * @return array Result of remove command
     */
    public function removeBulk(array $keys)
    {
        $result = array();
        foreach ($keys as $key) {
            $result[$key] = $this->_client->remove($this->_keyPrefix . $key);
        }

        return $result;
    }

    public function playScript($key, $script, $update = false)
    {
        return $this->_client->playScript($this->_keyPrefix . $key, $script, $update);
    }

    /**
     * Remove tag.
     *
     * @param mixed $tag
     * @access public
     * @return void
     */
    public function removeTag($tag)
    {
        throw new Exception('Not Implemented.');
    }

    /**
     * Clear all data.
     *
     * @access public
     * @return void
     */
    public function clear()
    {
        throw new Exception('Not Implemented.');
    }

    /**
     * Autoload class.
     *
     * @param  mixed $class
     * @access public
     * @return void
     */
    public static function autoload($className)
    {
        // Autoload class.
        // http://groups.google.com/group/php-standards/web/psr-0-final-proposal
        if (!class_exists($className, false)) {
            $className = ltrim($className, '\\');
            $fileName  = '';
            $namespace = '';
            if ($lastNsPos = strripos($className, '\\')) {
                $namespace = substr($className, 0, $lastNsPos);
                $className = substr($className, $lastNsPos + 1);
                $fileName  = str_replace('\\', DIRECTORY_SEPARATOR, $namespace) . DIRECTORY_SEPARATOR;
            }
            $fileName .= str_replace('_', DIRECTORY_SEPARATOR, $className) . '.php';

            require_once $fileName;
        }
    }
}


benchmarks/simple.php
<?php
/**
 * Add comment
 *
 * PHP version 5.3
 *
 * Copyright (c) 2011 Shinya Ohyanagi, All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *   * Neither the name of Shinya Ohyanagi nor the names of his
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * @category
 * @package
 * @version   $id$
 * @copyright (c) 2011 Shinya Ohyanagi
 * @author    Shinya Ohyanagi <sohyanagi@gmail.com>
 * @license   New BSD License
 */
error_reporting(E_ALL | E_STRICT);
$src = dirname(__DIR__) . DIRECTORY_SEPARATOR . 'src';
set_include_path(get_include_path()
    . PATH_SEPARATOR . $src
);
require_once 'Benchmark/Timer.php';


$hosts = array('localhost:8888');
$range = range(0, 1000);

echo 'OkuyamaClient' . PHP_EOL;
$t = new Benchmark_Timer();
$t->start();
$t->setMarker('Connect to Okuyama.');
require_once 'OkuyamaClient.class.php';
$client = new OkuyamaClient();
$client->setConnectionInfos($hosts);
$client->autoConnect();

$t->setMarker('Set 1000 items.');
foreach ($range as $v) {
    $client->setValue('key_' . $v, 'val_' . $v);
}
reset($range);
$t->setMarker('Get 1000 items.');
foreach ($range as $v) {
    $client->getValue('key_' . $v);
}
reset($range);
$t->setMarker('Remove 1000 items.');
foreach ($range as $v) {
    $client->removeValue('key_' . $v);
}
reset($range);
$t->setMarker('Close connection.');
$client->close();
$t->stop();
$t->display();

echo '\Net\Okuyama\Adapter\Socket' . PHP_EOL;
$t = new Benchmark_Timer();
$t->start();
$t->setMarker('Connect to Okuyama.');
require_once 'Net/Okuyama/Adapter.php';
require_once 'Net/Okuyama/Exception.php';
require_once 'Net/Okuyama/Adapter/Socket.php';
$client = new \Net\Okuyama\Adapter\Socket();
$client->autoConnect($hosts);
$t->setMarker('Set 1000 items.');
foreach ($range as $v) {
    $client->set('key_' . $v, 'val_' . $v);
}
reset($range);
$t->setMarker('Get 1000 items.');
foreach ($range as $v) {
    $client->get('key_' . $v);
}
reset($range);
$t->setMarker('Remove 1000 items.');
foreach ($range as $v) {
    $client->remove('key_' . $v);
}
reset($range);
$t->setMarker('Close connection.');
$client->close();
$t->stop();
$t->display();

echo '\Net\Okuyama' . PHP_EOL;
$t->start();
$t->setMarker('Connect to Okuyama.');
require_once 'Net/Okuyama.php';
$client = new \Net\Okuyama();
$client->connect($hosts);
$t->setMarker('Set 1000 items.');
foreach ($range as $v) {
    $client->set('key_' . $v, 'val_' . $v);
}
reset($range);
$t->setMarker('Get 1000 items.');
foreach ($range as $v) {
    $client->get('key_' . $v);
}
reset($range);
$t->setMarker('Remove 1000 items.');
foreach ($range as $v) {
    $client->remove('key_' . $v);
}
reset($range);
$t->setMarker('Close connection.');
$client->close();
$t->stop();
$t->display();
